### PR TITLE
Use LISTSXP instead of S4SXP as base object

### LINKED ISSUE #272: `dput(<R7_object>)` errors
reprex:
``` r
library(S7)
dput(new_class("Foo")())
#> Error in dput(new_class("Foo")()): cannot get a slot ("slots") from an object of type "NULL"
```

<sup>Created on 2023-01-05 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### LINKED ISSUE #281: Can't call debug() on a method
I thought this was a consequence of #279, but after installing #280, it appears to be a separate issue:

```r
library(S7)

dog <- new_class("dog", properties = list(
  name = class_character,
  age = class_numeric
))

speak <- new_generic("speak", "x")

method(speak, dog) <- function(x) {
  "Woof"
}

fn <- method(speak, dog)
debug(fn)
speak(dog())
```

This generates the error

```
debugging in: 
Error in (function (x) :
cannot get a slot ("slots") from an object of type "NULL"
```

It generated more messages before installing #280:
```
debugging in: Error in (function (x)  : 
  cannot get a slot ("slots") from an object of type "NULL"
Error during wrapup: badly formed function expression
Error: no more error handlers available (recursive errors?); invoking 'abort' restart
```

### PR DESCRIPTION
**Pull request by hadley**
Fixes #272. Fixes #281.

To do: 

* [x] add useful failure messages for internal generics (e.g. `sloop:::internal_generics()` and most common other S3 generics.
* [x] Add short-comings section which includes a note about future changes to underlying type.

``` r
is_s3_generic <- function(f)  {
  if (!is.function(f)) return(FALSE)
  
  uses <- codetools::findGlobals(f, merge = FALSE)$functions
  any(uses == "UseMethod")
}

funs <- mget(ls("package:base"), as.environment("package:base"))
names(Filter(is_s3_generic, funs))
#>  [1] "all.equal"                "anyDuplicated"           
#>  [3] "aperm"                    "as.array"                
#>  [5] "as.data.frame"            "as.Date"                 
#>  [7] "as.expression"            "as.function"             
#>  [9] "as.list"                  "as.matrix"               
#> [11] "as.null"                  "as.POSIXct"              
#> [13] "as.POSIXlt"               "as.single"               
#> [15] "as.table"                 "by"                      
#> [17] "chol"                     "close"                   
#> [19] "conditionCall"            "conditionMessage"        
#> [21] "cut"                      "determinant"             
#> [23] "diff"                     "droplevels"              
#> [25] "duplicated"               "flush"                   
#> [27] "format"                   "getDLLRegisteredRoutines"
#> [29] "is.na<-"                  "isSymmetric"             
#> [31] "julian"                   "kappa"                   
#> [33] "labels"                   "levels"                  
#> [35] "mean"                     "merge"                   
#> [37] "months"                   "mtfrm"                   
#> [39] "open"                     "plot"                    
#> [41] "pretty"                   "print"                   
#> [43] "qr"                       "quarters"                
#> [45] "rev"                      "row.names"               
#> [47] "row.names<-"              "rowsum"                  
#> [49] "scale"                    "seek"                    
#> [51] "seq"                      "sequence"                
#> [53] "solve"                    "sort"                    
#> [55] "split"                    "split<-"                 
#> [57] "subset"                   "summary"                 
#> [59] "t"                        "toString"                
#> [61] "transform"                "truncate"                
#> [63] "unique"                   "units"                   
#> [65] "units<-"                  "weekdays"                
#> [67] "with"                     "within"
```

<sup>Created on 2023-04-17 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### CODE REVIEW

File: src/method-dispatch.c [hunk 1]
```diff
@@ -89,7 +89,7 @@ SEXP S7_obj_dispatch(SEXP object) {
 }
 
 SEXP S7_object_() {
-  SEXP obj = PROTECT(Rf_allocSExp(S4SXP));
+  SEXP obj = PROTECT(Rf_allocSExp(LISTSXP));
```
Comment (DavisVaughan) [inline src/method-dispatch.c:92]: Not sure if this or `cons(R_NilValue, R_NilValue)` is "better"

**Comment by ltierney** [comment]
My suggestion would be

list1(install("S7_object"))

(except with the install done once and saved in a static
variable). That way unclass() would preserve some useful identifying
information.

luke

On Mon, 17 Apr 2023, Davis Vaughan wrote:

> 
> @DavisVaughan commented on this pull request.
> 
> ____________________________________________________________________________
> 
> In src/method-dispatch.c:
> 
> > @@ -89,7 +89,7 @@ SEXP S7_obj_dispatch(SEXP object) {
>  }
>
>  SEXP S7_object_() {
> -  SEXP obj = PROTECT(Rf_allocSExp(S4SXP));
> +  SEXP obj = PROTECT(Rf_allocSExp(LISTSXP));
> 
> Not sure if this or cons(R_NilValue, R_NilValue) is "better"
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you are subscribed to thisthread.[AA55UVGEVC2IKQBORFM5LWDXBUZRRA5CNFSM6AAAAAAW246FHSWGG33NNVSW45C7OR4
> XAZNRKB2WY3CSMVYXKZLTORJGK5TJMV32UY3PNVWWK3TUL5UWJTSSXIKWE.gif] Message ID:
> ***@***.***>
> 
> 
>

-- 
Luke Tierney
Ralph E. Wareham Professor of Mathematical Sciences
University of Iowa                  Phone:             319-335-3386
Department of Statistics and        Fax:               319-335-3017
    Actuarial Science
241 Schaeffer Hall                  email:   ***@***.***
Iowa City, IA 52242                 WWW:  http://www.stat.uiowa.edu

**Comment by hadley** [comment]
@ltierney @lawremi would love to get your feedback on the methods that cause S7 objects to fail cleanly when passed to S3 generics.

**Comment by mmaechler** [comment]
Tomas Kalinowski has proposed a patch towards using  new `OBJSXP` (as internal almost-synonym of `S4SXP`) as we had discussed (in person):    https://github.com/r-devel/r-svn/pull/128

I have tried a (reduced minimal) version of that sucessfully,  then also found that the `S7` package needed an update in order to stop working in such a patched version of R-devel, (the last commit being 6292876a62aa, keeping `S7` working in both "old" and "new" R (new not even in the R-devel code base yet ..)
I really think (possibly a version of)  that will be "the way" forward.      

IIUC we want the `LISTSXP` / `pairlist()` only just for an `S7` package  working with R <= 4.3.x ?
If yes, would we need to maintain two internally different branches of S7 for a while ?  What about serialization?
maybe something to discuss in  a meeting

**Comment by hadley** [comment]
@mmaechler did you see #303?

It looks like you committed your experiments directly into the main branch. Would you mind doing PRs in the future?

**Comment by mmaechler** [comment]
I think we have agreed to stay with `S4SXP` for current versions of R and hope to use `OBJSXP` in a future version of R,
as exemplified in the R-OBJSXP development branch  at https://svn.r-project.org/R/branches/R-OBJSXP/
(mirrored on gh here: https://github.com/wch/r-source/tree/R-OBJSXP).

Hence `LISTSXP` is no longer on the plate

**Pull request closed without merging**